### ISSUE TITLE
Recommended Reactant compile boundary for large fixed-corpus nearest-neighbor attention

### ISSUE THREAD

**Issue by AdityaPandeyCN** [EnzymeAD/Reactant.jl#2924]
Hello,

I'm working on tabular ML models in Julia and trying to use Reactant as the training/inference backend instead of Zygote.

Most models are standard tabular architectures, but one model is a nearest-neighbor / ModernNCA-style method. Conceptually:

1. Encode query batch `x`: `(d, B)`
2. Encode a fixed training corpus `cx`: `(d, N)`
3. Compute L2 distances `(N, B)`
4. Softmax over the corpus dimension
5. Predict as a weighted average of corpus targets

So eval/inference is like attention over the training set.

## What works

For small datasets, e.g. `N ~ 700`, Reactant works well. Training and eval with `@compile` complete successfully.

## What fails

For large tabular datasets, e.g. `N ~ 400_000` to `1_000_000`, compile can hang for a long time or the process exits without a Julia stack trace. The same workload with Zygote + GPU completes, slowly but stably.

## Relevant code

- ModernNCA implementation: https://github.com/AdityaPandeyCN/NeuroTabModels.jl/blob/emb-nca/src/models/ModernNCA/modernnca.jl
- Reactant extension: https://github.com/AdityaPandeyCN/NeuroTabModels.jl/blob/emb-nca/ext/NeuroTabModelsReactantExt.jl
- Eval callback compile boundary: https://github.com/AdityaPandeyCN/NeuroTabModels.jl/blob/emb-nca/src/Fit/callback.jl

## Question

What is the recommended Reactant pattern for this kind of model?

1. Should corpus encoding live outside `@compile`, with only query encoding + attention compiled?
2. Is dense `(N, B)` attention expected to compile/run for large fixed `N`, or should this be written as a tiled / streaming loop?

**Comment by ptiede** [comment]
Do you have an example of the code running? It is a little hard to figure out exactly what is `N` in this situation and how that maps to all the internal variables. 

All that said from the code you shared it does seem like you are going to unroll a massive loop which is usually the cause of long compilation times in my experience.

Specifically looking at

https://github.com/AdityaPandeyCN/NeuroTabModels.jl/blob/03042848d2ae6f0c6e0d4107d399388b3ccb897f/src/models/ModernNCA/modernnca.jl#L90-L99

and 

https://github.com/AdityaPandeyCN/NeuroTabModels.jl/blob/03042848d2ae6f0c6e0d4107d399388b3ccb897f/src/models/ModernNCA/modernnca.jl#L167-L182

are going to get unrolled without an `@trace`. I am not sure about the size of both loops as a function of `N` but since they are nested, the unroll will be substantial and lead to a massive code gen. Note the docs are alright at explaining this, admittedly but could use a pick of TLC https://enzymead.github.io/Reactant.jl/stable/tutorials/control-flow

If you want to figure out if increasing `N` leads to more and more code gen you can always pick `N=100` and `N=500` run `@code_hlo` instead of `@compile` and see if the number of lines it produces when going from `N=100` to `N=500`. That gives you a pretty good hint that Reactant is unrolling stuff internally. 

If this ends up being the source of the long compilation times the generic solution is to place an `@trace` on both the while and for loops to prevent unrolling.

**Comment by AdityaPandeyCN** [comment]
As you can see in one of the benchmark files [benchmarks/benchmark_mse.jl](https://github.com/AdityaPandeyCN/NeuroTabModels.jl/blob/03042848d2ae6f0c6e0d4107d399388b3ccb897f/benchmarks/benchmark_mse.jl), N corresponds to nobs:

```
nobs = Int(1e6)
num_feat = Int(100)
X = rand(Float32, nobs, num_feat)
dtrain = DataFrame(X, :auto)
```
So for that benchmark:

`N = nobs = 1,000,000`
ModernNCA then stores the corpus transposed:

`cx = permutedims(Matrix{Float32}(dtrain[:, feature_names]))`
So the shapes are:

```
X      = (1_000_000, 100)
dtrain = 1_000_000 rows
cx     = (100, 1_000_000)
N      = size(cx, 2) = 1_000_000
```
In eval/inference, the query batch is B = batchsize and attends over this full corpus, so the dense attention shape is:

`(N, B) = (1_000_000, 2048)`
That is the large N we are referring to.


Following your suggestion, I also tested this with `@code_hlo`.

For N=3000, the corpus encode loop is entered. After adding `@trace` and rewriting the loop body to use Reactant-compatible dynamic_slice / dynamic_update_slice, the HLO now contains a real:

`stablehlo.while`
with:

```
stablehlo.dynamic_slice
stablehlo.dynamic_update_slice
```
So the corpus encoding loop is no longer being unrolled.

However, even after fixing the loop unrolling, the large benchmark still hangs. The remaining large operation is the dense NCA attention itself:

```
kq = zk' * zq
softmax over (N, B)
```
which materializes:

`(N, B) = (1_000_000, 2048)`

**Comment by AdityaPandeyCN** [comment]
@avik-pal Corpus encode at large N is handled with `@trace` now. The remaining bottleneck is our ModernNCA eval attention(https://github.com/AdityaPandeyCN/NeuroTabModels.jl/blob/emb-nca/src/models/ModernNCA/modernnca.jl#L154-L165):
```
d = pairwise_dist(zq, zk) ./ temperature   # (N, B)
α = softmax(-d; dims=1)
pred = weighted_sum(α, cy)
```
with zk (d_emb, N), zq (d_emb, B) (e.g. N = 1e6, B = 2048). 
What is the recommended Reactant approach for this operation at that scale? Currently it just hangs and I have to kill the session.
